File: Object-C/SPPage/Controller/SPTabController.m
//
//  SPTabController.m
//  Radio
//
//  Created by sparrow on 16/01/2017.
//  Copyright © 2017 qzone. All rights reserved.
//

#import "SPTabController.h"
#import "SPTagBarScrollView.h"
#import "SPPageController.h"

#define SCREEN_WIDTH    [UIScreen mainScreen].bounds.size.width
#define SCREEN_HEIGHT   [UIScreen mainScreen].bounds.size.height

@interface SPTabController () <SPPageControllerDelegate, SPPageControllerDataSource, SPTabDataSource, SPTabDelegate>

@property (nonatomic, strong) UIView *tabView;

@property (nonatomic, strong) SPPageController *pageController;

@property (nonatomic, assign) CGFloat tabViewTop;

@property (nonatomic, assign) BOOL cannotScrollWithPageOffset;//为解决pagecontroller的横向滑动问题

@end

@implementation SPTabController


- (void)tabDragWithOffset:(CGFloat)offset
{
    
    [self setTabViewTop:[self tabScrollTopWithContentOffset:offset]];
}

- (CGFloat)tabScrollTopWithContentOffset:(CGFloat)offset
{
    CGFloat top = [self preferTabY]-offset;
    if (offset >= 0) {//上滑
        if (top <= self.minYPullUp) {
            top = self.minYPullUp;
        }
    } else {//下拉
        if (top >= self.maxYPullDown) {
            top = self.maxYPullDown;
        }
    }
    return top;
}


- (void)setTabViewTop:(CGFloat)tabViewTop
{
    self.tabView.frame = CGRectMake(self.tabView.frame.origin.x, tabViewTop, self.tabView.frame.size.width, self.tabView.frame.size.height);
}

- (CGFloat)tabViewTop
{
    return self.tabView.frame.origin.y;
}

- (void)reloadTab
{
    [self.tabView removeFromSuperview];
    self.tabView = nil;

    [self __setupTabBar];
}

- (void)pageviewControllerDidTransitiontoViewController:(UIViewController *)toVC fromVC:(UIViewController *)fromVC
{

    if ([fromVC conformsToProtocol:@protocol(SPPageSubControllerDataSource) ]) {
        UIViewController<SPPageSubControllerDataSource> *fromVCTemp =  (UIViewController<SPPageSubControllerDataSource> *)fromVC;
        fromVCTemp.preferScrollView.scrollsToTop = NO;
    }

    if ([toVC conformsToProtocol:@protocol(SPPageSubControllerDataSource) ]) {
        UIViewController<SPPageSubControllerDataSource> *toVCTemp =  (UIViewController<SPPageSubControllerDataSource> *)toVC;
        toVCTemp.preferScrollView.scrollsToTop = YES;
    }

    [self changeToSubControllerOffset:toVC isDelay:YES];

}

- (CGFloat)scrollViewOffsetAtIndex:(NSInteger)index
{
    return ([self preferTabY]- self.tabViewTop)- [self pageTopAtIndex:index];
}

- (void)changeToSubControllerOffset:(UIViewController *)toVC isDelay:(BOOL)isDelay
{
    if (!toVC || [self numberOfControllers] <=1) {
        self.cannotScrollWithPageOffset = NO;
        return;
    }
    
    if ([toVC conformsToProtocol:@protocol(SPPageSubControllerDataSource)]) {
        UIViewController<SPPageSubControllerDataSource> *toVCTemp =  (UIViewController<SPPageSubControllerDataSource> *)toVC;
        NSInteger newIndex = [self.pageController indexOfController:toVCTemp];
        CGFloat pageTop = [self pageTopAtIndex:newIndex];
        CGFloat top =  [self tabScrollTopWithContentOffset:[toVCTemp preferScrollView].contentOffset.y+pageTop];
        
        //如果计算出来的高度一样，不用去修改offset
        if ( fabs(top -self.tabViewTop) > 0.1) {
            CGFloat scrollOffset = [self scrollViewOffsetAtIndex:newIndex];
            self.cannotScrollWithPageOffset = NO;
            [toVCTemp preferScrollView].contentOffset =  CGPointMake(0, scrollOffset);
            
        } else {
            self.cannotScrollWithPageOffset = NO;
        }
        
    } else {
        self.cannotScrollWithPageOffset = NO;
    }
}



- (void)__setupPage
{
    self.pageController = [[SPPageController alloc] init];
    self.pageController.dataSource = self;
    self.pageController.delegate = self;
    [self.pageController updateCurrentIndex:[self preferPageFirstAtIndex]];
    self.pageController.view.frame = [self preferPageFrame];

    [self addChildViewController:self.pageController];
    [self.view addSubview:self.pageController.view];
    [self.pageController didMoveToParentViewController:self];

}

- (UIView *)customTabView
{
    return nil;
}

- (void)__setupTabBar
{
    if ([self preferSingleTabNotShow] && [self numberOfControllers] <=1) {

        return ;
    }

    UIView *tabView = [self customTabView];
    if (!tabView) {
        tabView = [[SPTagBarScrollView alloc] initWithFrame:[self preferTabFrame] dataSource:self delegate:self];

    }

    [self.view addSubview:tabView];
    self.tabView = tabView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self __setupData];
    [self __setupPage];
    [self __setupTabBar];
}

- (CGRect)preferTabFrame
{
    return CGRectMake([self preferTabX], [self preferTabY], [self preferTabW], [self preferTabHAtIndex:self.pageController.currentPageIndex]);
}

- (void)__setupData
{
    self.maxYPullDown = SCREEN_HEIGHT;
    self.minYPullUp = 64;//默认的navbar和statusbar的高度
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.cannotScrollWithPageOffset = NO;
    [self.pageController beginAppearanceTransition:YES animated:animated];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    [self.pageController endAppearanceTransition];
}


- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    self.cannotScrollWithPageOffset = YES;
    [self.pageController beginAppearanceTransition:NO animated:animated];

}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    [self.pageController endAppearanceTransition];
}

- (BOOL)preferSingleTabNotShow
{
    return NO;
}

- (void)reloadData
{
    [self reloadPage];
    [self reloadTab];
}

- (void)reloadPage
{
    [self.pageController updateCurrentIndex:[self preferPageFirstAtIndex]];
    self.pageController.view.frame = [self preferPageFrame];
    [self.pageController reloadPage];
}

- (void)updateTabBarWithIndex:(NSInteger)index
{
    if ([self.tabView isKindOfClass:[SPTagBarScrollView class]]) {
        SPTagBarScrollView *tabView = (SPTagBarScrollView *)self.tabView;
        [tabView reloadHighlightToIndex:index];
    }

}


- (void)reloadTabH:(BOOL)isTabScroll
{
    if (self.currentIndex < 0) {
        return;
    }
    self.tabView.frame  = [self preferTabFrame];

    if (!isTabScroll) {
        self.pageController.view.frame = [self preferPageFrame];
    } else {
        [self.pageController resizePageAtIndex:self.pageController.currentPageIndex offset:[self scrollViewOffsetAtIndex:self.currentIndex] isNeedChangeOffset:self.tabViewTop > self.minYPullUp atBeginOffsetChangeOffset:self.tabViewTop == self.minYPullUp];
    }
}

- (NSInteger)currentIndex
{
    return self.pageController.currentPageIndex;
}

#pragma SPPageControllerDelegate

-(void)scrollViewContentOffsetWithRatio:(CGFloat)ratio draging:(BOOL)draging
{
    if ([self.tabView isKindOfClass:[SPTagBarScrollView class]]) {

        if (!draging) {
            __weak SPTabController *wSelf = self;
            [UIView animateWithDuration:0.3 animations:^{
                SPTabController *bSelf = wSelf;
                SPTagBarScrollView *scrollView = (SPTagBarScrollView *)bSelf.tabView;
                [scrollView markViewScrollToIndex:ratio];
                scrollView.markViewScroll = NO;
            } completion:^(BOOL finished) {
                SPTabController *bSelf = wSelf;
                SPTagBarScrollView *scrollView = (SPTagBarScrollView *)bSelf.tabView;
                scrollView.markViewScroll = YES;
                [self updateTabBarWithIndex:floor(ratio+0.5)];

            }];
        } else {
            SPTagBarScrollView *scrollView = (SPTagBarScrollView *)self.tabView;
            [scrollView markViewScrollToContentRatio:ratio];
            [self updateTabBarWithIndex:floor(ratio+0.5)];
        }
    }

}

- (void)scrollWithPageOffset:(CGFloat)offset index:(NSInteger)index
{
    [self tabDragWithOffset:offset+[self pageTopAtIndex:index]];
}


- (void)pageviewController:(SPPageController*)pageController willTransitionFromVC:(UIViewController *)fromVC toViewController:(UIViewController *)toVC
{

    [self changeToSubControllerOffset:toVC isDelay:NO];
}

- (void)pageviewController:(SPPageController*)pageController didTransitionFromVC:(UIViewController *)fromVC toViewController:(UIViewController *)toVC
{
    [self pageviewControllerDidTransitiontoViewController:toVC fromVC:fromVC];
    if ([self.tabView isKindOfClass:[SPTagBarScrollView class]]) {
        SPTagBarScrollView *scrollView = (SPTagBarScrollView *)self.tabView;
        [scrollView scrollTagToIndex:self.pageController.currentPageIndex];

    }

}

- (void)pageviewController:(SPPageController*)pageController willLeaveFromVC:(UIViewController *)fromVC toViewController:(UIViewController *)toVC
{
    [self changeToSubControllerOffset:toVC isDelay:NO];
}

- (void)pageviewController:(SPPageController*)pageController didLeaveFromVC:(UIViewController *)fromVC toViewController:(UIViewController *)toVC
{
    [self pageviewControllerDidTransitiontoViewController:toVC fromVC:fromVC];
}

#pragma SPPageControllerDataSource

- (NSInteger)numberOfControllers
{
    return 0;
}

- (CGRect)preferPageFrame
{
    return CGRectMake(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
}

- (UIViewController *)controllerAtIndex:(NSInteger)index
{
    return nil;
}

- (CGFloat)pageTopAtIndex:(NSInteger)index
{
    return [self preferTabY] + [self preferTabHAtIndex:index]-[self preferPageFrame].origin.y;
}

- (UIScreenEdgePanGestureRecognizer *)screenEdgePanGestureRecognizer
{
    UIScreenEdgePanGestureRecognizer *screenEdgePanGestureRecognizer = nil;
    if (self.navigationController.view.gestureRecognizers.count > 0)
    {
        for (UIGestureRecognizer *recognizer in self.navigationController.view.gestureRecognizers)
        {
            if ([recognizer isKindOfClass:[UIScreenEdgePanGestureRecognizer class]])
            {
                screenEdgePanGestureRecognizer = (UIScreenEdgePanGestureRecognizer *)recognizer;
                break;
            }
        }
    }

    return screenEdgePanGestureRecognizer;
}

- (BOOL)isPreLoad
{
    return NO;
}

#pragma SPTabDataSource 规范化的TAB设计样式可以放在这里

- (NSString *)titleForIndex:(NSInteger)index
{
    return nil;
}

- (CGFloat)preferTabHAtIndex:(NSInteger)index
{
    return 40;
}

- (CGFloat)preferTabY
{
    return KNAVIGATIONANDSTATUSBARHEIGHT;
}

- (CGFloat)preferTabX
{
    return 0;
}

- (CGFloat)preferTabW
{
    return SCREEN_WIDTH;
}

-(NSInteger)preferPageFirstAtIndex
{
    return 0;
}

- (NSInteger)numberOfTab
{
    return [self numberOfControllers];
}


-(CGFloat)tabTopForIndex:(NSInteger)index
{
    return 0;
}

- (CGFloat)tabWidthForIndex:(NSInteger)index
{
    NSString *text = [self titleForIndex:index];
    if (text.length <=3) {
        return 73;
    } else {
        return 105;
    }
}

- (UIColor *)titleColorForIndex:(NSInteger)index
{
    return [UIColor blackColor];
}

- (UIColor *)titleHighlightColorForIndex:(NSInteger)index
{
    return [UIColor orangeColor];
}

- (UIFont *)titleFontForIndex:(NSInteger)index
{
    return [UIFont systemFontOfSize:13.0];
}

- (UIColor *)tabBackgroundColor
{
    return [UIColor whiteColor];
}

- (NSInteger)preferTabIndex{
    return [self preferPageFirstAtIndex];
}

- (BOOL)needMarkView
{
    return YES;
}

- (CGFloat)markViewBottom
{
    return [self preferTabHAtIndex:self.pageController.currentPageIndex]-7;
}

-(CGFloat)markViewWidthForIndex:(NSInteger)index
{
    NSString *text = [self titleForIndex:index];

    CGRect rect = [text boundingRectWithSize:CGSizeMake(20000, 40) options:NSStringDrawingUsesFontLeading|NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName:[self titleFontForIndex:index ]} context:nil];

    return rect.size.width;
}

-(UIColor *)markViewColorForIndex:(NSInteger)index
{
    return [UIColor orangeColor];
}

#pragma SPTabDelegate
- (void)didPressTabForIndex:(NSInteger)index
{
    if ([self isSubPageCanScrollForIndex:index]) {
        [self.pageController showPageAtIndex:index animated:YES];

    }
    
}

- (void)willChangeInit
{
    self.cannotScrollWithPageOffset = YES;
}

-(BOOL)isTabCanPressForIndex:(NSInteger)index
{
    return [self isSubPageCanScrollForIndex:index];
}

-(BOOL)isSubPageCanScrollForIndex:(NSInteger)index
{
    return YES;
}


@end

File: Swift/SPPage/core/SPTabController.swift
//
//  SPTabController.swift
//  SPPage
//
//  Created by GodDan on 2018/3/11.
//  Copyright © 2018年 GodDan. All rights reserved.
//

import UIKit

class SPTabController : UIViewController,SPPageControllerDelegate,SPPageControllerDataSource,SPTabDelegate,SPTabDataSource {
    
    var maxYPullDown:CGFloat = kScreenHeight
    var minYPullUp: CGFloat = 64
    
    private var tabView:UIView?
    private var pageVC:SPPageController!
    private var _cannotScrollWithPageOffset:Bool = false
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setupPage()
        self.setupTabBar()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self._cannotScrollWithPageOffset = false
        self.pageVC.beginAppearanceTransition(true, animated: animated)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.pageVC.endAppearanceTransition()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self._cannotScrollWithPageOffset = true
        self.pageVC.beginAppearanceTransition(false, animated: animated)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        self.pageVC.endAppearanceTransition()
    }
    
    func preferSingleTabNotShow() -> Bool {
        return false
    }
    
    func customView() -> UIView? {
        return nil
    }
    
    func reloadData() {
        self.reloadPage()
        self.reloadTab()
    }
    
    private func reloadTab() {
        self.tabView?.removeFromSuperview()
        self.tabView = nil
        self.setupTabBar()
    }
    
    private func reloadPage() {
        self.pageVC.view.frame = self.preferPageFrame()
        self.pageVC.reloadPage()
    }
    
    private func setupPage() {
        self.pageVC = SPPageController()
        self.pageVC.dataSource = self
        self.pageVC.delegate = self
        
        self.pageVC.view.frame = self.preferPageFrame()
        self.addChildViewController(self.pageVC)
        self.view.addSubview(self.pageVC.view)
        self.pageVC.didMove(toParentViewController: self)
        
    }
    
    private func setupTabBar() {
        if self.preferSingleTabNotShow() && self.numberOfControllers() <= 1 {
            return
        }
        
        self.tabView = self.customView() ?? SPTagBarScrollView(frame: self.preferTabFrame(index: self.pageVC.currentPageIndex), tabDelagate: self, tabDataSource: self)
    
        self.view.addSubview(self.tabView!)
    }
    
    private func scrollViewOffset(index:NSInteger) -> CGFloat {
        return self.preferTabFrame(index: index).origin.y - (self.tabView?.frame.origin.y)! - self.pageTopAtIndex(index:index)
    }
    
    private func tabDragWithOffset(offset:CGFloat) {
        self.tabView?.frame.origin.y = self.tabScrollTopWithContentOffset(offset: offset)
    }
    
    
    func tabScrollTopWithContentOffset(offset:CGFloat) -> CGFloat {
        var top:CGFloat = self.preferTabFrame(index: self.pageVC.currentPageIndex).origin.y - offset
        if offset >=  0 {
            if top <= self.minYPullUp {
                top = self.minYPullUp
            }
        } else {
            if top >= self.maxYPullDown {
                top = self.maxYPullDown
            }
        }
        
        return top
    }
    
    private func changeToSubControllerOffset(toVc:UIViewController?, isDelay:Bool) {
        if toVc == nil || self.numberOfControllers() <= 1 {
            self._cannotScrollWithPageOffset = false
            return
        }
        
        if toVc is SPPageSubControllerDataSource {
            if let tempVC = toVc as? SPPageSubControllerDataSource {
                let newIndex =  self.pageVC.indexOf(vc:tempVC as! UIViewController)
                let pageTop = self.pageTopAtIndex(index: newIndex)
                let top = self.tabScrollTopWithContentOffset(offset: (tempVC.preferScrollView()?.contentOffset.y)!+pageTop)
                if fabs(top-(self.tabView?.frame.origin.y)!) > 0.1 {
                    let scrollOffset = self.scrollViewOffset(index: newIndex)
                    self._cannotScrollWithPageOffset = false
                    tempVC.preferScrollView()?.contentOffset.y = scrollOffset
                } else {
                    self._cannotScrollWithPageOffset = false
                }
                
            }
        } else {
            self._cannotScrollWithPageOffset = false

        }
    }
    
    // SPPageDataSource
    
    func pageTopAtIndex(index: NSInteger) -> CGFloat {
        return self.preferTabFrame(index: index).origin.y + self.preferTabFrame(index: index).size.height -  self.preferPageFrame().origin.y
    }
    
    func screenEdgePanGestureRecognizer() -> UIScreenEdgePanGestureRecognizer? {
        
        if let gestureRecognizers = self.navigationController?.view.gestureRecognizers {
            for recognizer:UIGestureRecognizer? in gestureRecognizers {
                if recognizer is UIScreenEdgePanGestureRecognizer {
                    return recognizer as? UIScreenEdgePanGestureRecognizer
                }
            }
        }
      
        return nil
    }
    
    func preferPageFirstAtIndex() -> NSInteger {
        
        return 0
    }
    
    func controller(index: NSInteger) -> UIViewController {
        return UIViewController.init()
    }
    
    func numberOfControllers() -> NSInteger {
        return 0
    }
    
    func preferPageFrame() -> CGRect {
        return CGRect.init(x: 0, y: 0, width: kScreenWidth, height: kScreenHeight)
    }
    
    func isPreLoad() -> Bool {
        return false
    }

    func isSubPageCanScrollForIndex(index: NSInteger) -> Bool {
        return true
    }
    
    // SPTabDataSource
    func title(index: NSInteger) -> String {
        return ""
    }
    
    func numberOfTab() -> NSInteger {
        return self.numberOfControllers()
    }
    
    func preferTabFrame(index: NSInteger) -> CGRect {
        return CGRect.init(x: 0, y: 0, width: kScreenWidth, height: 40)
    }
    
    func tabTopForIndex(index: NSInteger) -> CGFloat {
        return 0
    }
    
    func tabWidth(index: NSInteger) -> CGFloat {
        if self.title(index: index).count <= 3 {
            return 73
        } else {
            return 105
        }
    }
    
    func titleColor(index: NSInteger) -> UIColor {
        return UIColor.black
    }
    
    func titleHighlightColor(index: NSInteger) -> UIColor {
        return UIColor.orange
    }
    
    func titleFont(index: NSInteger) -> UIFont {
        return UIFont.systemFont(ofSize: 13.0)
    }
    
    func tabBackgroundColor() -> UIColor {
        return UIColor.white
    }
    
    func preferTabIndex() -> NSInteger {
        return self.preferPageFirstAtIndex()
    }
    
    func needMarkView() -> Bool {
        return true
    }
    
    func markViewBottom(index: NSInteger) -> CGFloat {
        return self.preferTabFrame(index: self.pageVC.currentPageIndex).size.height - 7
    }
    
    func markViewColor(index: NSInteger) -> UIColor {
        return UIColor.orange
    }
    
    func markViewWidth(index: NSInteger) -> CGFloat {
        let text = NSString (string: self.title(index: index))
        let rect = text.boundingRect(with: CGSize.init(width: 20000, height: 40), options: [.usesFontLeading,.usesLineFragmentOrigin] , attributes: [NSAttributedStringKey.font:self.titleFont(index: index)], context: nil)
        return rect.size.width
    }
    
    func tabCanPress(index: NSInteger) -> Bool {
        return true
    }
    
    // SPTabDelegate
    func didPressTab(index: NSInteger) {
        if self.isSubPageCanScrollForIndex(index: index) {
            self.pageVC.showPage(index: index, animated: true)
        }
    }
    
    //SPPageControllerDelegate
    func willChangeInit() {
        self._cannotScrollWithPageOffset = true
    }
    
    func scrollViewContentOffsetWithRatio(ratio: CGFloat, draging: Bool) {
        if self.tabView is SPTagBarScrollView {
            if !draging {
                UIView.animate(withDuration: 0.3, animations: {[weak self] in
                    if let weakSelf = self {
                        if let tabView = weakSelf.tabView as?  SPTagBarScrollView {
                            tabView.markViewScroll(index: NSInteger(ratio))
                            tabView.markViewScroll = false
                        }
                        
                    }
                }, completion: { [weak self](finished:Bool) in
                    if let weakSelf = self {
                        if let tabView = weakSelf.tabView as?  SPTagBarScrollView {
                            tabView.markViewScroll = true
                            tabView.reloadHighlight(index: NSInteger(floor(ratio+0.5)))
                        }
                    }
                })
            } else {
                if let tabView = self.tabView as?  SPTagBarScrollView {
                    tabView.markViewScroll(contentRatio: Double(ratio))
                    tabView.reloadHighlight(index: NSInteger(floor(ratio+0.5)))
                }
            }
        }
    }
    
    func scrollWithPageOffset(pageOffset: CGFloat, index: NSInteger) {
        self.tabDragWithOffset(offset: pageOffset + self.pageTopAtIndex(index: index))
    }
    
    func pageviewControllerWillLeave(pageVC: SPPageController, fromVC: UIViewController, toVC: UIViewController) {
        self.changeToSubControllerOffset(toVc: toVC, isDelay: false)
    }
    
    func pageviewControllerDidLeave(pageVC: SPPageController, fromVC: UIViewController, toVC: UIViewController) {
        self.pageviewControllerDidChange(fromVC: fromVC, toVC: toVC)

    }
    
    func pageviewControllerWillTransition(pageVC: SPPageController, fromVC: UIViewController, toVC: UIViewController) {
        self.changeToSubControllerOffset(toVc: toVC, isDelay: false)
    }
    
    func pageviewControllerDidTransition(pageVC: SPPageController, fromVC: UIViewController, toVC: UIViewController) {
        self.pageviewControllerDidChange(fromVC: fromVC, toVC: toVC)
        if let tempView = self.tabView as? SPTagBarScrollView {
            tempView.scrollTab(index: self.pageVC.currentPageIndex)
        }
    }
    
    func pageviewControllerDidChange(fromVC: UIViewController, toVC: UIViewController) {
        if let tempVC = fromVC as? SPPageSubControllerDataSource {
            tempVC.preferScrollView()?.scrollsToTop = false
        }
        
        if let tempVC = toVC as? SPPageSubControllerDataSource {
            tempVC.preferScrollView()?.scrollsToTop = false
        }
        
        self.changeToSubControllerOffset(toVc: toVC, isDelay: true)
    }
    
    func cannotScrollWithPageOffset() -> Bool {
        return self._cannotScrollWithPageOffset
    }
}
